daliuge-engine/test/test_drop.py

```diff
@@ -747,8 +753,106 @@ def checkDropStates(aStatus, dStatus, eStatus, lastByte):
         checkDropStates(
             DROPStates.COMPLETED, DROPStates.COMPLETED, DROPStates.COMPLETED, b"k"
         )
+        self.assertEqual(b"ejk", droputils.allDropContents(d))
+
+    def test_objectAsNormalAndAsyncStreamingInput(self):
+        """
+        A test that checks that a DROP can act as normal and async streaming input
+        of different AppDROPs at the same time. We use the following graph:
+
+        A --|--> B --> D
+            |--> C --> E
+
+        Here B uses A as a streaming input, while C uses it as a normal
+        input
+        """
+
+        class DropAsyncStream(AsyncIterator):
+            _d = Deque()
+            _end = False
+            async def __anext__(self):
+                while True:
+                    if len(self._d) > 0:
+                        data = self._d.popleft()
+                        return data
+                    elif self._end:
+                        raise StopAsyncIteration
+                    else:
+                        await asyncio.sleep(0)
+
+            def append(self, data):
+                self._d.append(data)
+
+            def end(self):
+                self._end = True
 
+        class LastCharWriterApp(AppDROP):
+            _lastByte = None
+            _stream = DropAsyncStream()
+
+            def run(self):
+                asyncio.run(self.arun())
+
+            async def arun(self):
+                async for data in self._stream:
+                    outputDrop = self.outputs[0]
+                    self._lastByte = data[-1:]
+                    outputDrop.write(self._lastByte)
+
+            def dataWritten(self, uid, data):
+                self._stream.append(data)
+
+            def dropCompleted(self, uid, drop_state):
+                self._stream.end()
```

Comment:
here's a different approach to  stream objects using the existing streaming drops. This works with python threading but multiprocessing has a few more catches